src/Interceptor/HeaderInterface.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/**
+ * This file is part of Temporal package.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Temporal\Interceptor;
+
+use IteratorAggregate;
+use Temporal\Api\Common\V1\Header;
+use Temporal\DataConverter\DataConverterInterface;
+use Temporal\DataConverter\Type;
+
+/**
+ * @psalm-type TKey=array-key
+ * @psalm-type TValue=string
+ * @psalm-import-type TypeEnum from Type
+ *
+ * @extends IteratorAggregate<TKey, string>
+ */
+interface HeaderInterface extends \Countable, IteratorAggregate
+{
+    /**
+     * Checks if any value present.
+     */
+    public function isEmpty(): bool;
+
+    /**
+     * @param TKey $index
+     * @param Type|TypeEnum|mixed $type
+     *
+     * @return mixed Returns {@see null} if value not found.
+     */
+    public function getValue(int|string $index, mixed $type = null): mixed;
+
+    /**
+     * @param TKey $key
+     * @param TValue $value
+     *
+     * @psalm-mutation-free
+     */
+    public function withValue(int|string $key, string $value): self;
+
+    /**
+     * Make a protobuf Header message.
+     *
+     * @return Header
+     */
+    public function toHeader(): Header;
+
+    /**
+     * @param DataConverterInterface $converter
+     */
+    public function setDataConverter(DataConverterInterface $converter);
```

Comment:
It's odd that you can set a data converter specifically for headers. We typically allow data converters to be set on clients and/or workers, and sometimes around a specific command in a workflow, but it's odd to have it publicly exposed that a converter could be set for a specific header

Reply:
In this PR I'll mark it `@internal` and I might remove it in the next PR.

The way it's done now is that headers are not decoded if they're not requested, and are sent as is if they haven't been changed. And those values that were added through `withValue()` are encoded into Payload before sending.

I'll try to do it differently: headers on input will be decoded immediately and packed again only before sending.

OK?

Reply:
Yeah, I think that makes good sense